src/pgduckdb_ddl.cpp

```diff
@@ -176,65 +275,143 @@ duckdb_drop_table_trigger(PG_FUNCTION_ARGS) {
+				 * their transaction lifecycles.
+				 */
+				PreventInTransactionBlock(true, "DuckDB queries");
+				connection = pgduckdb::DuckDBManager::CreateConnection();
+				pgduckdb::DuckDBQueryOrThrow(*connection, "BEGIN TRANSACTION");
+			}
+			HeapTuple tuple = SPI_tuptable->vals[proc];
+
+			char *postgres_schema_name = SPI_getvalue(tuple, SPI_tuptable->tupdesc, 1);
+			char *table_name = SPI_getvalue(tuple, SPI_tuptable->tupdesc, 2);
+			char *drop_query = psprintf("DROP TABLE %s.%s", pgduckdb_db_and_schema(postgres_schema_name, true),
+			                            quote_identifier(table_name));
+			pgduckdb::DuckDBQueryOrThrow(*connection, drop_query);
+		}
 	}
 
 	/*
-	 * For now, we don't support DuckDB queries in transactions. To support
-	 * write queries in transactions we'll need to link Postgres and DuckdB
-	 * their transaction lifecycles.
+	 * And now we basically do the same thing as above, but for TEMP tables.
+	 * For those we need to check our in-memory temporary_duckdb_tables set.
 	 */
-	PreventInTransactionBlock(true, "DuckDB queries");
-
-	auto connection = pgduckdb::DuckDBManager::CreateConnection();
+	ret = SPI_exec(R"(
+		SELECT objid, object_name
+		FROM pg_catalog.pg_event_trigger_dropped_objects()
+		WHERE object_type = 'table'
+			AND schema_name = 'pg_temp'
+		)",
+	               0);
 
-	pgduckdb::DuckDBQueryOrThrow(*connection, "BEGIN TRANSACTION");
+	if (ret != SPI_OK_SELECT)
+		elog(ERROR, "SPI_exec failed: error code %s", SPI_result_code_string(ret));
 
 	for (auto proc = 0; proc < SPI_processed; ++proc) {
 		HeapTuple tuple = SPI_tuptable->vals[proc];
 
-		char *object_identity = SPI_getvalue(tuple, SPI_tuptable->tupdesc, 1);
-		// TODO: Handle std::string creation in a safe way for allocation failures
-		pgduckdb::DuckDBQueryOrThrow(*connection, "DROP TABLE IF EXISTS " + std::string(object_identity));
+		bool isnull;
+		Datum relid_datum = SPI_getbinval(tuple, SPI_tuptable->tupdesc, 1, &isnull);
+		if (isnull) {
+			elog(ERROR, "Expected relid to be returned, but found NULL");
+		}
+		Oid relid = DatumGetObjectId(relid_datum);
+		if (temporary_duckdb_tables.count(relid) == 0) {
+			/* It was a regular temporary table, so this DROP is allowed */
+			continue;
+		}
+		if (!connection) {
+			/*
+			 * For now, we don't support DuckDB queries in transactions. To support
+			 * write queries in transactions we'll need to link Postgres and DuckdB
+			 * their transaction lifecycles.
+			 */
+			PreventInTransactionBlock(true, "DuckDB queries");
+			connection = pgduckdb::DuckDBManager::CreateConnection();
+			pgduckdb::DuckDBQueryOrThrow(*connection, "BEGIN TRANSACTION");
+		}
+		char *table_name = SPI_getvalue(tuple, SPI_tuptable->tupdesc, 2);
+		pgduckdb::DuckDBQueryOrThrow(*connection,
+		                             std::string("DROP TABLE pg_temp.main.") + quote_identifier(table_name));
+		temporary_duckdb_tables.erase(relid);
 	}
 
+	AtEOXact_GUC(false, save_nestlevel);
```

Comment:
Is that what restore the GUC (search_path/duckdb.execution)? If so we might have an issue when an exception is thrown no? (either C++ or PG)

Reply:
At transaction end the postgres transaction management will automatically rollback these changes too, even on failure. This call is only necessary on success, because we want any further code to use the old values for these GUCs again.